snakemake/executors.py

```diff
@@ -2401,25 +2408,67 @@ def _generate_job_resources(self, job):
         resources = {"regions": self.regions, "virtualMachine": virtual_machine}
         return resources
 
+    def _set_workflow_sources(self, warning_size_gb=20):
+        """We only add files from the working directory that are config related
+           (e.g., the Snakefile or a config.yml equivalent), or checked into git. 
+           Additionally, given that we encourage these packages to be small,
+           we set a warning at 20GB.
+        """
+
+        def check_size(f):
+            """A helper function to check the filesize, and return the file
+               to the calling function
+            """
+            gb = bytesto(os.stat(f).st_size, "g")
+            if gb > warning_size_gb:
+                logger.warning(
+                    "File {} (size {} GB) is greater than the {} GB suggested size "
+                    "Consider uploading larger files to storage first.".format(
+                        f, gb, warning_size_gb
+                    )
+                )
+            return f
+
+        self.workflow_sources = []
+
+        for wfs in self.workflow.get_sources():
+            if os.path.isdir(wfs):
+                for (dirpath, dirnames, filenames) in os.walk(wfs):
+                    self.workflow_sources.extend(
+                        [check_size(os.path.join(dirpath, f)) for f in filenames]
+                    )
+            else:
+                self.workflow_sources.append(check_size(os.path.abspath(wfs)))
+
     def _generate_build_source_package(self):
         """in order for the instance to access the working directory in storage,
            we need to upload it. This file is cleaned up at the end of the run.
            We do this, and then obtain from the instance and extract.
         """
+        # Add the execution scripts
+        # import code
+        # code.interact(local=locals())
+
         # We will generate a tar.gz package, renamed by hash
         tmpname = next(tempfile._get_candidate_names())
         targz = os.path.join(tempfile.gettempdir(), "snakemake-%s.tar.gz" % tmpname)
         tar = tarfile.open(targz, "w:gz")
 
-        # Add all files to a package
-        for root, dirs, files in os.walk(self.workdir):
+        # Add all workflow_sources files
+        for filename in self.workflow_sources:
+            arcname = filename.replace(self.workflow.basedir + os.path.sep, "")
+            tar.add(filename, arcname=arcname)
+
+        # Add snakemake hidden directory
+        for root, dirs, files in os.walk(os.path.join(self.workdir, ".snakemake")):
```

Comment:
I don't yet understand what this block is needed for...

Reply:
Why do we need to upload the hidden directory?

Reply:
We have to add the entrypoint scripts, which are located in there.

Reply:
And this is also fixed (meaning that the .snakemake directory is not uploaded, and we generate the package once).